diff --git a/scripts/hbsAgent.service b/scripts/hbsAgent.service
index bd4bcd6..9f8790c 100644
--- a/scripts/hbsAgent.service
+++ b/scripts/hbsAgent.service
@@ -5,9 +5,9 @@ Before=pmon.service
 
 [Service]
 Type=forking
-ExecStart=/etc/rc.d/init.d/hbsAgent start
-ExecStop=/etc/rc.d/init.d/hbsAgent stop
-PIDFile=/var/run/hbsAgent.pid
+ExecStart=/etc/init.d/hbsAgent start
+ExecStop=/etc/init.d/hbsAgent stop
+PIDFile=/run/hbsAgent.pid
 
 [Install]
 WantedBy=multi-user.target
